src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,253 @@
+        WebUtils.toHttp(response).sendError(HttpServletResponse.SC_FORBIDDEN,
+          errorMessage);
+        return false;
+      }
+    }
+  }
+
+
+  private boolean checkGen3Authorization(String teamProjectRole, String login) throws Exception {
+    logger.debug("Checking Gen3 Authorization for 'team project'={} and user={} using service={}", teamProjectRole, login, this.authorizationUrl);
+    RestTemplate restTemplate = new RestTemplate();
+    String arboristAuthorizationURL = this.authorizationUrl;
+    String requestBody = String.format("{\"username\": \"%s\"}", login);
+    String jsonResponseString = restTemplate.postForObject(arboristAuthorizationURL, requestBody, String.class);
+
+    JSONObject jsonObject = new JSONObject(jsonResponseString);
+
+    if (!jsonObject.keySet().contains(teamProjectRole)) {
+      logger.warn("User is not authorized to access this team project's data");
+      return false;
+    } else {
+      JSONArray teamProjectAuthorizations = jsonObject.getJSONArray(teamProjectRole);
+      logger.debug("Found authorizations={}", teamProjectAuthorizations);
+      // We expect only one authorization rule per teamproject:
+      if (teamProjectAuthorizations.length() != 1) {
+        logger.error("Only one authorization rule expected for 'teamproject'={}, found={}", teamProjectRole,
+          teamProjectAuthorizations.length());
+        return false;
+      }
+      JSONObject teamProjectAuthorization = teamProjectAuthorizations.getJSONObject(0);
+
+      // check if the authorization contains the right "service" and "method" values:
+      String expectedMethod = "access";
+      String expectedService = "atlas-argo-wrapper-and-cohort-middleware"; // TODO - make the service name configurable?
+      String service = teamProjectAuthorization.getString("service");
+      String method = teamProjectAuthorization.getString("method");
+      logger.debug("Parsed service={} and method={}", service, method);
+      if (!method.equalsIgnoreCase(expectedMethod)) {
+        logger.error("The 'teamproject' authorization method should be '{}', but found '{}'", expectedMethod, method);
+        return false;
+      }
+      logger.debug("Parsed method is as expected");
+      if (!service.equalsIgnoreCase(expectedService)) {
+        logger.error("The 'teamproject' authorization service should be '{}', but found '{}'", expectedService, service);
+        return false;
+      }
+      logger.debug("Parsed service is as expected");
+      return true;
+    }
+  }
+
+  /**
+   * Tries to find the team project information in the given request or as part of the
+   * stored this.authorizer.getCurrentTeamProjectRoleForCurrentUser(). Returns null
+   * if nothing can be found.
+   */
+  private String extractTeamProjectFromRequestParameters(ServletRequest request) throws Exception {
+    // Get the url
+    HttpServletRequest httpRequest = (HttpServletRequest) request;
+    String url = httpRequest.getRequestURL().toString();
```

Comment:
Is this where you allow for the authorization header to hold the team project string? That would probably address things.

Reply:
see previous comments. Please let me know if there is still an issue.

Reply:
I don't think it is an issue per your discussion.